### DISCUSSION TITLE
Primary Key changed from Auto Increment to UUID - Directus Files and Users

### DISCUSSION THREAD

**Discussion by philleepflorence** [directus/directus#4050] [Archive]
### Questions
Is there any reason for replacing the **Primary Key** Field of **Directus Files** or **Directus Users** from INT-11 to CHAR-36?
What prevents adding a **Field** and **Index** for UUID?
If a cause for the breaking change, is there any method migrate from D8 - INT to D9 - UUID?

### Background
A compatibility test failed when comparing the core collections in D8 and D9 - paraphrasing the result below:
1. Setting Primary Key **Index** via **CHAR** on **InnoDB** failed optimization requirement.
2. Triggers and Stored Procedures requiring Auto Increment Prediction, MIN, and MAX internal functions failed.

### Context 
1. Some existing projects managed by D8 are very large - 100+ collections (tables), 100M+ items (rows) with many more O2M (FK) relationships to Directus Files and Directus Users - and require strict optimization especially the size of the index used to reference the row.
2. Some functions need to be installed within (or are not permitted outside) **MySQL**, **MariaDB**, (or potentially new clients running on **OracleDB**) as opposed to using **Hooks** outside of the DB. Also, accessing the next possible ID, as well as MIN, MAX, is when using Auto Increment

### Suggestion
A less destructive and more agnostic approach - to support data imports and fresh installs.
```
CREATE TABLE collection (
  id INT NOT NULL AUTO_INCREMENT,
  uuid CHAR NOT NULL,
  .....
  PRIMARY KEY(id),
  UNIQUE(uuid)
);
```

### References
[Universal Unique Identifier - UUID](https://oracle-base.com/articles/9i/uuid-9i)
[Mysql 8.0: UUID support](https://mysqlserverteam.com/mysql-8-0-uuid-support/)
[MySQL InnoDB Primary Key Choice: GUID/UUID vs Integer Insert Performance](https://kccoder.com/mysql/uuid-vs-int-insert-performance/)
[https://blog.programster.org/mysql-performance-when-using-uuid-for-primary-key](https://blog.programster.org/mysql-performance-when-using-uuid-for-primary-key)
[MySQL UUID Smackdown: UUID vs. INT for Primary Key](https://www.mysqltutorial.org/mysql-uuid/)

**Comment by rijkvanzanten** [comment]
> Is there any reason for replacing the Primary Key Field of Directus Files or Directus Users from INT-11 to CHAR-36?

Security. It's easy to guess files/users when the IDs are auto-incremented

> What prevents adding a Field and Index for UUID?

Nothing?

> If a cause for the breaking change, is there any method migrate from D8 - INT to D9 - UUID?

Recreate the files through the API, replacing the foreign keys with the updated IDs

**Comment by philleepflorence** [comment]
After auditing, it looks like most of the issue would be disk space as the User and File UUID would be used as FK in many collections.

> Security. It's easy to guess files/users when the IDs are auto-incremented

Noted...User and File IDs were never used by the web applications though.

> Recreate the files through the API, replacing the foreign keys with the updated IDs

Hopefully, the internal SQL UUID functions should suffice, though it would be a great task to convert the larger projects should they move to D9.
